### DISCUSSION TITLE
How to run an external program in a target directory

### DISCUSSION THREAD

**Discussion by nstgc** [janet-lang/janet#1415] [Q&A]
I want to run an external program in several different directories in parallel. In Clojure, to run an external program in a given directory, I would run `(sh "pwd" :dir "some/path")` and get back `some/path`. How would this be done in Janet?

**Comment by ianthehenry** [comment]
This isn't really an answer, but I don't know of a nice way to do this right now.

Neither `os/spawn` nor https://github.com/andrewchambers/janet-posix-spawn support `posix_spawn_file_actions_addchdir_np` or which is kinda what you want. You can `os/cd` before you spawn each subprocess if you don't care about mangling your current working directory (or are willing to restore it).

I think it would probably be reasonable to add `posix_spawn_file_actions_addchdir_np` to `os/spawn`? But I don't know what platforms it's actually supported on / what that `ifdef` would look like.

https://github.com/andrewchambers/janet-sh is a nice subprocess DSL but it doesn't have a way to spawn subprocesses without synchronously awaiting them (you can do this with the janet-posix-spawn library it's based on; I don't *think* you can asynchronously await the native `os/spawn` but I could be wrong and it seems like you *should* be able to). It might be reasonable to add a `$&` for direct control over the spawned process... but it would be better if janet-sh just used the event loop so you could use `ev/spawn` to run multiple subprocesses concurrently for free.

**Reply by nstgc** [reply]
Ah, okay. I figured parrelel `cd`s would be one of those side effects that can really mess up a script.

I do suppose that in this case I could just run a separate instance in each directory, then another script to tie it all together afterwards.

Thanks. it's hard keeping up with all the new stuff in Janet sometimes.
